### ISSUE TITLE
Consider where to put debug tools

### ISSUE THREAD

**Issue by thesunny** [ianstormtaylor/slate#2810]
#### Do you want to request a _feature_ or report a _bug_?

feature

#### What's the current behavior?

There's a bunch of debug tools that are essential to working with Slate (especially Android for me).

I've added them as plugins that are enabled by `debug` settings. They aren't performance issues as they aren't added unless they are specifically set in the `debug` settings but they do add to the size of the bundle.

#### What's the expected behavior?

@ianstormtaylor I think this is something more for you. Here's a few things that I think can make this better and some options:

1. A way to add plugins that go before the stack (In front of the `before` plugin). Some options are `<Editor beforePlugins={myBeforePlugins} />` (works only for that instance), or `Editor.beforePlugins = myBeforePlugins` for global or `Editor.setBeforePlugins(myBeforePlugins)`. Kind of leaning to `<Editor beforePlugins={myBeforePlugins} />`

2. Also need a way to get plugins out of Slate so basically where they should exist like `import DebugPlugin from 'slate/debug-plugin'` or `import DebugPlugin from 'slate/plugins/debug'`. I do suggest we keep this in Slate proper though because the tools need to be compatible with the current version of Slate to be useful IMO.

**Comment by ianstormtaylor** [comment]
I think having each package have it's own `debug` information sounds good to me, so I'm down to keep them. Specifically the current debug plugin in `slate-react` sounds good to me. I think it might be good to simplify it a bit, potentially eliminating the batch events, and making the stringification only happen when `debug.enabled === true` or whatever for performance.

Ideally in a future iteration of middleware we'd be able to use functional composition better such that these plugins aren't even in the stack if not enabled... unsure.

**Comment by thesunny** [comment]
@ianstormtaylor One step ahead on this one. :)

The code checks `debug.enabled` and the debug plugins are not in the stack if it's not enabled.

WRT batch events, I started writing them in one plugin with the regular debug events but found that it complicated the code without doing much to reduce the total amount of code. I made two different ones as I found that the use cases for each were quite different although batch events are mostly useful for debugging Android only. If the MutationObserver works out though, batch events might not be that valuable in which case I'll remove them.
